### PR TITLE
feat: return ControlFlow from AST visitor methods

### LINKED ISSUE #14: Return `ControlFlow` from AST visitor methods
Currently AST visitors do not return anything, whereas it would be beneficial to return a `ControlFlow` value, with the associated break type defined by implementors. By default, all functions should always return `Continue(())`, and implementors may choose to break early.

https://github.com/paradigmxyz/sulk/blob/5f827dd5a0fed12f8e57c48ac60e371b98d921a3/crates/ast/src/visit.rs#L7-L10

### PR DESCRIPTION
**Pull request by mahmudsudo**
closes #14

### CODE REVIEW

File: crates/ast/src/visit.rs [hunk 1]
```diff
@@ -4,139 +4,154 @@ use crate::ast::*;
...
-        fn visit_source_unit(&mut self, source_unit: &#mut SourceUnit<'ast>) {
+        fn visit_source_unit(&mut self, source_unit: &#mut SourceUnit<'ast>) -> ControlFlow<()> {
             // TODO: SAFETY: Idk
             let source_unit = unsafe { trustme::decouple_lt #_mut(source_unit) };
             let SourceUnit { items } = source_unit;
             for item in items.iter #_mut() {
-                self.visit_item #_mut(item);
+                if let ControlFlow::Break(_) = self.visit_item #_mut(item) {
+                    return ControlFlow::Break(());
+                }
```
Comment (DaniPopes) [inline crates/ast/src/visit.rs]: This can be `?`

File: crates/sema/src/ast_passes.rs [hunk 2]
```diff
@@ -107,11 +109,14 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
             }
             _ => {}
         }
+        ControlFlow::Continue(())
     }
 
-    // Intentionally override unused default implementations to reduce bloat.
```
Comment (DaniPopes) [inline crates/sema/src/ast_passes.rs:112]: Please keep this comment

File: crates/ast/src/visit.rs [hunk 3]
```diff
@@ -4,139 +4,154 @@ use crate::ast::*;
 use solar_data_structures::trustme;
 use solar_interface::{Ident, Span};
 use solar_macros::declare_visitors;
+use std::ops::ControlFlow;
 
 declare_visitors! {
     /// AST traversal.
     pub trait Visit VisitMut <'ast> {
-        fn visit_source_unit(&mut self, source_unit: &#mut SourceUnit<'ast>) {
+        fn visit_source_unit(&mut self, source_unit: &#mut SourceUnit<'ast>) -> ControlFlow<()> {
```
Comment (DaniPopes) [inline crates/ast/src/visit.rs]: We should let implementors define the `Break` type; it should be possible to add `type BreakValue;` in the trait and these return types to be `ControlFlow<Self::BreakValue>`.

One can specify `Never` as the `BreakValue`, this should be in the docs for the type

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/ast/src/visit.rs (hunk 1 above)
Comment (mahmudsudo) [reply crates/ast/src/visit.rs]: putting a ``?`` would prompt an error in the if let statement

**Review by DaniPopes — APPROVED** [review]
Thanks!

**Pull request merged**